sumpy/tools.py

```diff
@@ -976,19 +981,198 @@ def wait(self):
+                assignee=x[(*broadcast_dims, i2)],
+                expression=0,
+                id=f"reset_{ilev}",
+                depends_on=frozenset([f"copy_{ilev}"])),
+            lp.Assignment(
+                assignee=table_idx,
+                expression=nfft*iN1_sum*(iN2 + N2*iN1),
+                id=f"idx_{ilev}",
+                depends_on=frozenset([f"reset_{ilev}"]),
+                temp_var_type=lp.Optional(np.uint32)),
+            lp.Assignment(
+                assignee=exp,
+                expression=exp_table[table_idx % n],
+                id=f"exp_{ilev}",
+                depends_on=frozenset([f"idx_{ilev}"]),
+                within_inames=frozenset(map(lambda x: x.name,
+                    [*broadcast_dims, iN1_sum, iN1, iN2])),
+                temp_var_type=lp.Optional(complex_dtype)),
+            lp.Assignment(
+                assignee=x[(*broadcast_dims, ifft + nfft * (iN1*N2 + iN2))],
+                expression=(x[(*broadcast_dims, ifft + nfft*(iN1*N2 + iN2))]
+                    + exp * temp[ifft + nfft * (iN2*N1 + iN1_sum)]),
+                id=f"update_{ilev}",
+                depends_on=frozenset([f"exp_{ilev}"])),
+        ]
+
+        domains += [
+            f"[ifft_{ilev}]: 0<=ifft_{ilev}<{nfft}",
+            f"[iN1_{ilev}]: 0<=iN1_{ilev}<{N1}",
+            f"[iN1_sum_{ilev}]: 0<=iN1_sum_{ilev}<{N1}",
+            f"[iN2_{ilev}]: 0<=iN2_{ilev}<{N2}",
+            f"[i_{ilev}]: 0<=i_{ilev}<{n}",
+            f"[i2_{ilev}]: 0<=i2_{ilev}<{n}",
+        ]
+
+    for idom, dom in enumerate(domains):
+        if not dom.startswith("{"):
+            domains[idom] = "{" + dom + "}"
+
+    if n == 1:
+        insns += ["x[:, i] = x[:, i] {dup=i}"]
+    elif inverse:
+        insns += [
+            lp.Assignment(
+                assignee=x[(*broadcast_dims, i3)],
+                expression=x[(*broadcast_dims, i3)] / n,
+                depends_on=frozenset([f"update_{len(factors) - 1}"]),
+            ),
+        ]
+
+    if name is None:
+        if inverse:
+            name = f"ifft_{n}"
+        else:
+            name = f"fft_{n}"
+
+    knl = lp.make_kernel(domains, insns,
+        kernel_data=[
+            lp.GlobalArg("x", shape=shape, is_input=False, is_output=True,
+                dtype=complex_dtype),
+            lp.GlobalArg("y", shape=shape, is_input=True, is_output=False,
+                dtype=complex_dtype),
+            lp.TemporaryVariable("exp_table", shape=(n,),
+                dtype=complex_dtype),
+            lp.TemporaryVariable("temp", shape=(n,),
+                dtype=complex_dtype),
+            ...],
+        name=name,
+        fixed_parameters=fixed_parameters,
+        lang_version=lp.MOST_RECENT_LANGUAGE_VERSION,
+        index_dtype=index_dtype,
+    )
+    if broadcast_dims:
+        knl = lp.split_iname(knl, "j0", 32, inner_tag="l.0", outer_tag="g.0")
+        knl = lp.add_inames_for_unused_hw_axes(knl)
+
+    return knl
+
+
+def _get_fft_backend(queue):
```

Comment:
Make this return an enum for clarity? Maybe type-annotate?

Reply:
Fixed in 5cd2b73